### PR TITLE
BLOGS-1347 Legacy Routes

### PR DESCRIPTION
**Pull request by mattgreenham**
This is the first part of our 'legacy' blogs handling. 

These routes redirect the old url structure to the new. Existing can be seen [here](https://github.com/bbc/blogs/blob/master/webapp/php/lib/BBC/Blogs4/Action/Helper/RedirectLegacyPaths.php)

Have changed iStats page type implementation as not all Controllers which extend the base should set iStats labels.

### CODE REVIEW

**Comment by programmes-developers** [comment]
Can one of the admins verify this patch?

File: src/Controller/LegacyRoutesRedirectController.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        return $this->generateGuid($namespace);
+    }
+
+    private function generateAuthorGuid(string $blogId, string $authorName): GUID
+    {
+        $namespace = '/' . $blogId . '/authors/' . $authorName;
+        return $this->generateGuid($namespace);
+    }
+
+    private function generateGuid(string $namespace): GUID
```
Comment (votemike) [inline src/Controller/LegacyRoutesRedirectController.php]: A test for this?

File: src/Controller/BaseController.php [hunk 2]
```diff
@@ -66,7 +70,15 @@ public function __construct()
         $this->response()->headers->set('X-Content-Type-Options', 'nosniff');
     }
 
-    abstract protected function getIstatsPageType(): string;
+    protected function setIstatsPageType(string $pageType)
+    {
+        $this->istatsPageType = $pageType;
+    }
+
+    protected function getIstatsPageType(): string
```
Comment (votemike) [inline src/Controller/BaseController.php]: Is this used anymore?

File: config/routes.yaml [hunk 3]
```diff
@@ -84,3 +84,7 @@ feed_rss_tags:
     path: /blogs/{blogId}/tags/{tagId}/rss
     defaults: { _controller: App\Controller\TagRssFeedController }
     requirements: { blogId: '[a-z0-9]+', tagId: '[A-Za-z0-9-]+'}
+
+# Include legacy routes, this should be the last thing in the file
```
Comment (BPScott) [inline config/routes.yaml]: Symfony 4's new kernel shall load routes from the `config/routes` folder without having to explicitly include them in other config files. (see https://github.com/bbc/blogs-frontend/blob/master/src/Kernel.php#L55)

If you rename `config/legacy_routes.yaml` to `config/routes/legacy.yaml` then you can remove the modifications to this file

File: src/Controller/BaseController.php [hunk 4]
```diff
@@ -134,6 +141,17 @@ protected function setLocale(string $locale)
         $this->locale = $locale;
     }
 
+    protected function cachedRedirect($url, $status = 302): RedirectResponse
+    {
+        $headers = $this->response->headers->all();
+        return new RedirectResponse($url, $status, $headers);
+    }
+
+    protected function cachedRedirectToRoute($route, array $parameters = array(), $status = 302): RedirectResponse
```
Comment (votemike) [inline src/Controller/BaseController.php]: Tiny thing, `[]` instead of `array()`?

File: src/Controller/LegacyRoutesRedirectController.php [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        parent::__construct();
+
+        // Cache these redirects for an hour
+        $this->response()->setPublic()->setMaxAge(3600);
+    }
+
+    public function redirectPostShow(string $blogId, string $postName)
+    {
+        $guid = $this->generatePostGuid($blogId, $postName);
+        return $this->cachedRedirectToRoute('post', ['blogId' => $blogId, 'guid' => $guid], 301);
```
Comment (votemike) [inline src/Controller/LegacyRoutesRedirectController.php]: Do people prefer `301` or `Response::HTTP_MOVED_PERMANENTLY`? I think the 2nd is more descriptive if you can't remember each HTTP code off the top of your head.

Comment (Tasiobg) [reply src/Controller/LegacyRoutesRedirectController.php]: In most cases like in this one I would advice to use constants

**Pull request merged**